Optical Tactile Sensor Plugin

mcres opened the pull request:
**Optical Tactile Sensor Plugin for GSoC 2020.**

It retrieves the contacts given by a contact sensor and computes the normal surfaces of an object given the data returned by a Depth Camera. It also visualizes these forces in a similar way as Gazebo Classic, so this PR partially addresses #112.

Future work will be focused on:

- Tutorial explaining the plugin.
- Computation performance.
- Merging the data returned by contact and camera sensors.

![out2](https://user-images.githubusercontent.com/14854080/87320244-fe104980-c52a-11ea-9696-07d5fb824874.png)

mcres commented:
CC @mabelzhang

azeey commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePlugin::Update(const UpdateInfo &_info,
```
We should document this somewhere, but in general, we don't recommend using `System::Update` call unless absolutely necessary. This is the phase where physics is running and the contact points may not be ready yet. Consider putting this in `PostUpdate`.

mabelzhang replied:
Thanks! I was not aware of that.

chapulina replied:
> We should document this somewhere,

There's a bit [here](https://ignitionrobotics.org/api/gazebo/3.0/createsystemplugins.html):

_Generally, systems that do not manage physics stepping will not need to implement the ISystemUpdate interface._

And the [System](https://ignitionrobotics.org/api/gazebo/3.0/classignition_1_1gazebo_1_1System.html) API docs also mention:

_Update (...) Used for physics simulation step_

But we can always make it clearer.

mabelzhang replied:
I will admit we did find that page (I see it in our Slack records), but I didn't read it closely... oops.

mcres replied:
My bad, I was aware of that documentation as well. I would suggest to specify what exactly _physics simulation step_ is or writing an example, so it's easy to know for newcomers if they should implement it. I'm not sure what it is. 

Anyway, moving to `PostUpdate`. Thanks!

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::UnpackPointCloudMsg(
```
Is it necessary to subscribe to the point cloud message? Processing a point cloud is very time-consuming. Are you using the point cloud just to get the XYZ in the world frame? You should be able to calculate this using the camera matrix. It's a lot faster to do the matrix calculation rather than looping through the point cloud. Matrix speed >> for-loop speed.

mcres replied:
In `UnpackPointCloudMsg` I am getting the XYZ point in the camera frame corresponding to each (i,j) pixel, out of the message returned by the Depth Camera.

Maybe you are referring to `ComputeNormalForces` method?
Hmm, not sure how to do that, is there any example that implements that?

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+  int columnCounter = 0;
+
+  int index = 1;
+
```
Extra empty line

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+  bool visualizeColumn = true;
+  int columnCounter = 0;
+
+  int index = 1;
```
A more descriptive variable name for the index, since there's a lot of indexing going on here.

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+    }
+    visualizeColumn = true;
+    visualizeRow = (this->resolution == ++rowCounter);
+      if (visualizeRow)
```
indentation

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+            ->mutable_orientation()->set_z(forceMarkerWorldPose.Rot().Z());
+          this->forceMarkerMsg.mutable_pose()
+            ->mutable_orientation()->set_w(forceMarkerWorldPose.Rot().W());
+
```
Extra blank line

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+
+      // Downsampling for visualization
+      if (visualizeRow && visualizeColumn)
+        {
```
indentation

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+
+  // Visualization is downsampled according to plugin parameter <resolution>
+  // if <visualize_forces> is set to true
+  bool visualizeRow = true;
```
Maybe this can be simplified to `visualizeRow = _visualizeForces;`?

mcres replied:
Currently it couldn't, as I keep incrementing `columnCounter` and `rowCounter` even if there's no visualization. I could fix this with some more `if` statements. This wouldn't simplify the method, but I guess that avoiding to increment those variables is worth it. What do you think?

mabelzhang replied:
Hmm are they being incremented? Looks like `if (!_visualizeForces)`, the `continue` will skip the increment lines.

mcres replied:
I totally agree, it's quite mixed up currently. 
Is it possible to span the same class into more than one file? I'm thinking about defining the visualization methods into two separate `.cc` and `.hh` files.

mabelzhang replied:
I think you can split into more than one file. For example, [`multicopter_control` plugin does this.](https://github.com/ignitionrobotics/ign-gazebo/tree/master/src/systems/multicopter_control).

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+      // todo(anyone) multiply vector by contact forces info
+      normalForces[i][j] = direction.Normalized();
+
+      if (!_visualizeForces)
```
I guess right now there isn't a reason to compute the normals at all if visualizeForce == false. You can add a todo that the normals will be published. That can be in a later PR as this one is already pretty long.

mcres commented on examples/worlds/optical_tactile_sensor_plugin.sdf:
```diff
@@ -0,0 +1,272 @@
+            </clip>
+          </camera>
+        </sensor>
+        <sensor name="optical_tactile_sensor" type="contact">
```
You're right.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+  /// \param[in] _visualizeForces Whether to visualize the forces or not
+  ///
+  /// Implementation inspired by
+  /// https://stackoverflow.com/questions/
```
FYI, I saw [this method](https://ignitionrobotics.org/api/math/6.4/classignition_1_1math_1_1Quaternion.html#ab8d157cebd35fcc2a4420c1cb4949b36) in the API and just decided to do the same.

azeey replied:
I don't know if we have a general policy about this, but I think it's good to cite it. At the very least, it serves as some documentation as to what the code is doing.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+    // Get sdf parameters
+    if (!_sdf->HasElement("resolution"))
+    {
+        ignerr << "Missing required parameter <resolution>." << std::endl;
```
Great, I'll make `<visualize_forces>` optional as well.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+               const bool _visualizeForces);
+
+  /// \brief Resolution of the sensor in pixels to skip.
+  public: int resolution{100};
```
Fair point! However, I'll keep the `int` type and check for positive values. If the user specifies a negative value in the `.sdf` file, `_sdf->Get<uint64_t>("resolution")` would convert it to some really big positive number. Besides, seems that [Google Style Guide](https://google.github.io/styleguide/cppguide.html#Integer_Types) doesn't recommend using unsigned types, so that's another good reason.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+void OpticalTactilePluginPrivate::InterpolateData(
+                    const ignition::msgs::Contact &contact)
+{
+    if (contact.position_size() == 4)
```
Totally agree!

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+                            EntityComponentManager &_ecm,
+                            EventManager &)
+{
+    this->dataPtr->sdfConfig = _sdf->Clone();
```
Yes, it does. I understand that this is more efficient, but ~9/10 plugins use the `Clone()` way. Maybe it is not safe having two pointers pointing to the same variable? Besides, the purpose of `sdfConfig` is to read the `sdf` in the `Load()` function, so it's only read.

azeey replied:
`Clone` is used because `sdf::Element::GetElement` is not a const function. If you don't need `GetElement`, you can use `_sdf` directly.

azeey commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+
+    // Configure subscriber for Depth Camera images
+    sdf::ElementPtr depthCameraSdf =
+      _sdf->GetParent()->GetElement("link")->GetElement("sensor");
```
This seems like it's looking for the `<sensor>` tag in the first `<link>` element. What if the sensor is in a different link? You probably want to either specify the sensor name in the plugin or look for entities that have either `components::DepthCamera` or `components::RgbdCamera`.

mcres replied:
So the plugin is supposed to have one link only, with one depth camera and one contact sensor. Otherwise, it wouldn't make sense. I should still be able to differentiate between those two kind of sensors, because they could be placed in different order. Right now it takes the first sensor it finds.

Apart from that, maybe I'm missing checks on how the `sdf` model is built (one link with two specific sensors)?

azeey replied:
I see. I didn't know about the one link assumption. I was thinking that a user can add this system to any model that has a depth camera sensor. I'm good with keeping it simple for now and adding more configuration options later. I would suggest adding a TODO about it though.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,682 @@
+  public: ignition::msgs::PointCloudPacked cameraMsg;
+
+  /// \brief 3D vector to store unpacked XYZ points
+  std::vector<std::vector<std::vector<float>>> imageXYZ;
```
I had tried `ignition::common::Image` but [it is not](https://github.com/ignitionrobotics/ign-common/blob/master/graphics/src/Image.cc#L191) compatible with the data format of `ignition::msgs::PointCloudPacked`. 
What I can do is work with a raw pointer and somehow map the (i, j, XYZ) to the pointer position (basically what I'm doing [here](https://github.com/mcres/ign-gazebo/blob/optical_tactile_plugin/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L565)). That way I would get rid of `UnpackPointCloudMsg()` and I'd just run `ComputeNormalForces()`.

mabelzhang replied:
That sounds about right. The file has changed quite a bit since the comment and I'm not seeing the line you meant. Could you point to the line again?

@azeey Is there something equivalent to Eigen::MatrixXf that can be used to manipulate a matrix of size width * height * 3, where width and height are passed in by the user? Or does it need to be a raw array with manual indexing?

mcres replied:
Yup, here's the updated link: https://github.com/mcres/ign-gazebo/blob/optical_tactile_plugin/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L807. Instead of the `imageXYZ[i][j][k]` vector when computing `dxdi` and `dxdj`, there would be a mapping function instead, for instance `imageXYZ(i,j,k)`.

azeey replied:
You should be able to use Eigen since ign-gazebo depends on it indirectly via the eigen3 component of ign-math6. We already use it in the multicopter_control system. If you do use Eigen, be sure to specify the appropriate [storage option](https://eigen.tuxfamily.org/dox/group__TopicStorageOrders.html) (RowMajor or ColumnMajor) to improve performance

mcres replied:
It seems that we have two options here:

1. Using MatrixXf: I would have to unpack the data anyways but it would be more efficient than a nested vector (I assume?).
2. Manually indexing the raw array returned in the camera msg: I wouldn't have to unpack the data, but I would have to call the previously mentioned indexing function each time I want to access the value.

I'd go with no. 2, what do you think?

azeey replied:
@mabelzhang, I think I was too quick to answer about `MatrixXf` because I now realize that you really want a ndarray/tensor, which exists in Eigen, but is unsupported. @mcres, option 2 makes sense. I don't think you can get more efficient than calculating the index into the raw array in the point cloud message. Copying them into a new 3d std::vector incurs allocation costs and the allocated memory is unlikely to be contiguous between dimensions, which would result in unnecessary cache misses.

mcres commented:
Sorry, I'll push changes to another branch for a future PR!
I'll take a look to those indentations as well.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,927 @@
+    }
+
+    if (!depthCameraInLink || !contactSensorInLink ||
+      sensorsInsideLink.size() != 2)
```
I agree on the fact that the user should have that choice. However, if they have either +1 contact sensors or depth cameras, currently I don't have a way to know which one I should use for the plugin. I guess I could use the `name` attribute?

mabelzhang replied:
I was thinking if they wanted to have some other sensor in the link, like an IMU or something completely different from what this plugin deals with. Looks like `sensorsInsideLink` is counting not just depth and contact but all sensors. I think what you want is to only count depth and contact, then you would keep a counter that increments within the two if-statements above this, as opposed to checking the number of sensors in the link.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,927 @@
+        uint64_t steps2 =
+            contact1.Distance(contact4) / this->resolution;
+
+        for (uint64_t index1 = 0; index1 <= steps1; ++index1)
```
> Unconventional practice to use 0 with `<=`

Just for curiosity, should it be `0` with `<` and `1` with `<=`?

> Right now, the loop produces (steps1+1) * steps2 iterations.

Error there

mabelzhang replied:
Yes usually `0` with `<` is the preference (unless you're in MATLAB, which is......... another story).

mcres replied:
Ok, thanks!

> Error there

I think I was too quick saying that. The for loop actually computes `(steps1+1) * (steps2+1)` points that way. If you take for instance a 3x3 cm with a 1 cm resolution you get 4x4 points that way. I do `index1 <= steps1` because otherwise I don't get the last column. Does that makes sense?

mabelzhang replied:
Does it compute steps2+1 in the inner loop? That was my original question. Currently, it looks like it's only computing steps2 iterations, in `[0, steps2)`. The outer loop is computing `[0, steps1]`. So the total right now is (steps1+1) * steps2, I think.

mcres replied:
It computes `steps2` in the inner loop and `steps1 + 1` in the outer loop, but the thing is that the total number of steps is not the same that the total number of points `interpolatedVector` passes through

mabelzhang replied:
So to clarify, why does the inner loop only needs `steps2` iterations? Does it not also need `steps2 + 1`, to get the last row?

mcres replied:
No, it doesn't.

mcres replied:
Because one of the rows is already being got [here](https://github.com/ignitionrobotics/ign-gazebo/blob/16732f6f7b7a1e36f743fed07458296bafa177c1/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L409). The reason of that if-block statement is to not make `interpolatedVector` go outside of the rectangle drawn by the contacts' data.

mabelzhang replied:
Hmm I think a few comments in this loop, at least a comment at the top of this loop noting what the indexing is doing, and changing the variable names `1` and `2` to something referring to row and column like `r` and `c`, would make this method more readable. Or is it not known whether it's row or column because the directions might change?
Are the 4 points returned in clockwise / counterclockwise order? A comment on that where contacts 1-4 are declared also helps the reader.

I just realized that this is a method that I said elsewhere that we should remove / comment out? I can't find that spot anymore. Is this that method? If we're going to remove it anyway, then we don't need to spend time on it.

mcres replied:
> I just realized that this is a method that I said elsewhere that we should remove / comment out? I can't find that spot anymore. Is this that method? If we're going to remove it anyway, then we don't need to spend time on it.

Yeah, here's the comment https://github.com/ignitionrobotics/ign-gazebo/pull/229#discussion_r466915015

mabelzhang replied:
Yeah I saw that, but the line number and method name there didn't match with this version, so I was confused. Now I see that that comment is from the 20th commit, and this set is only the first 19, so that comment wasn't showing here.

So here are the two options. I would say just remove this method since we're not going to use it anymore. You can always retrieve it from Git history if you need it later. If you really want to keep it now, then it should be tidied up with additional comments as mentioned above, and then the method call commented out.

mcres replied:
Removed the method in e216bb2

mcres commented:
Apart from the feedback modifications, I've made quite some refactoring based on the feedback as well, but hopefully it's not too confusing.
The main changes are the following:

- As provided in the feedback, I've separated the forces visualization from the computation. For that purpose, I've added `Visualization.hh` and `Visualization.cc`.
- Depth camera callback: Instead of storing the point cloud into a nested `std::vector` variable, I get a specific point from the raw pointer returned when I compute the normal forces, so that should be faster.
- The `Configure` and `Load` methods had overlapping functionality before. Now I basically use `Configure` to read all of the sdf parameters and `Load` to check the plugin structure with the `EntityComponentManager` and some additional things.
- I've removed some attributes/methods that I wrote at the beginning of the project but actually don't use/need them.
- In `OpticalTactilePlugin.cc`, I've changed the order in which the methods are written, so hopefully the file is more intuitive to read and things are easier to find.
- Indentation!

mabelzhang commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
 
-      // todo(mcres) normal forces will be published even if visualization
-      // is turned off
+      // todo(mcres) Normal forces will be computed even if visualization
```
Replace `will be` with something more clear. `should be` (i.e. and are not)? Or `are`?

mcres replied:
e216bb2

mcres commented on src/systems/optical_tactile_plugin/Visualization.cc:
```diff
@@ -0,0 +1,239 @@
+  positionMarkerWorldPose.Correct();
+
+  // Set markers pose messages from previously computed poses
+  this->forceMarkerMsg.mutable_pose()->mutable_position()->set_x(
```
The problem is that `msgs::Set()` are [implemented](https://ignitionrobotics.org/api/msgs/5.1/Utility_8hh.html) for `Vector3d`, and `Pose3d` and I am using `float`. `Color` can be changed though.

mabelzhang replied:
Float can always be automatically promoted to double. It should let you do it.

mcres replied:
I meant I'm using `Vector3f`, `Quaternionf` and `Pose3f` (float) instead of `Vector3d`, `Quaterniond` and `Pose3d` (double), so I can't use `msgs::Set()` between those types

mabelzhang replied:
The variables like `forceMarkerSensorPosition` in this method are all local scope, and you can use `Vector3d` etc for them. That will not take up much extra space since it's only a few variables, and they only exist for the duration of this method. With the messages being in double, you won't be saving much space anyway. That way, the following blocks will be a lot cleaner.

mcres replied:
e216bb2

I've `msgs::Set` the `Color` msgs, and for the remaining `Pose3d` msg needed I wrote [this](https://github.com/ignitionrobotics/ign-gazebo/blob/e216bb2132d0a77ca77dc31a50435487273fbe09/src/systems/optical_tactile_plugin/Visualization.cc#L79-L83). [Same](https://github.com/ignitionrobotics/ign-gazebo/blob/e216bb2132d0a77ca77dc31a50435487273fbe09/src/systems/optical_tactile_plugin/Visualization.cc#L181-L191) for the `Vector3d` msgs.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+    for (const auto &contact : csd->Data().contact())
     {
-      this->cameraUpdateRate = depthCameraSdf->Get<float>("update_rate");
+      this->dataPtr->objectCollisionEntity =
```
That's actually used for publishing the messages, which is not in the scope of this PR: https://github.com/mcres/ign-gazebo/blob/optical_tactile_plugin_2/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L648

If it's not a problem I'd like to merge it that way

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
 
-    // Get collisions from the sensor
-    for (const Entity &entity : _entities)
+  if (contacts)
```
Do you say that because `InterpolateContactData` is not publishing the contacts?
Should I write a comment explaining why is it commented out?

mabelzhang replied:
I think we said we don't need to interpolate anymore, because it is limited to a box shape, and we only did it as a placeholder before we had depth camera images, right? I don't remember if we decided something else. I think we only need to keep the actual contacts from the physics engine, and not the ones interpolated, because they can be very off if the contacts are on a mesh.

mcres replied:
Removed code: https://github.com/ignitionrobotics/ign-gazebo/pull/229#discussion_r469805612

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
-void OpticalTactilePlugin::PostUpdate(
-    const ignition::gazebo::UpdateInfo &_info,
-    const ignition::gazebo::EntityComponentManager &_ecm)
+void OpticalTactilePluginPrivate::Enable(const bool _value)
```
Already implemented here: https://github.com/mcres/ign-gazebo/blob/optical_tactile_plugin_2/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L613-L626

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
-void OpticalTactilePluginPrivate::UnpackPointCloudMsg(
-                        const ignition::msgs::PointCloudPacked &_msg)
+ignition::math::Vector3f OpticalTactilePluginPrivate::MapPointCloudData(
+  const u_int32_t &_i, const uint32_t &_j)
```
Nope, error there.

mabelzhang replied:
Hmm I don't understand what you mean when you say error there. Do you mean compiler error?

mcres replied:
Sorry, I meant it was a typo.

mcres replied:
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
-{
-    IGN_PROFILE("TouchPluginPrivate::PreUpdate");
+  // Store depth camera update rate and offset from model
+  auto offset = depthCameraSdf->Get<ignition::math::Pose3d>("pose");
```
If not specified, it gives you a `(0, 0, 0, 0, 0, 0)` pose. I think this is well documented in the tutorial and in the example world, so if the user skips the `<pose>` tag I should store that value.

mabelzhang replied:
Right. I am saying this should not be a required tag, as is implied by the following if-block, which errors and returns if it's not specified. It should be kept optional and the identity be used if it's not specified.

mcres replied:
Oh I see, but I'm checking the `<update_rate>` tag in that block.

mabelzhang replied:
Oh oops sorry I misread.

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
+  int fieldIndex = 0;
+
+  // X coordinate
+  measuredPoint.X() = *reinterpret_cast<float *>(
```
Yeah we should check that.
I understand that you suggest checking that the first time we receive a message from the depth camera? Or is it every time? Probably the former is enough.

mabelzhang replied:
Yeah just once should be sufficient, to save time as much as possible. Presumably the user would not have two nodes both publishing depth images to the same topic - they can, but that is not usually recommended as it would mess up the topic, timing, etc.

mcres replied:
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
-      if (visualizeColumn)
-          columnCounter = 0;
+      markerPosition = this->MapPointCloudData(i, j);
+      this->visualizePtr->VisualizeNormalForce(i, j, markerPosition,
```
About `/list`, check this comment: https://github.com/ignitionrobotics/ign-gazebo/pull/250#discussion_r467732754

Looking at what the `/marker` service [does](https://github.com/ignitionrobotics/ign-gazebo/blob/master/src/rendering/MarkerManager.cc#L592-L597), it shouldn't be very difficult to implement that.

Anyway, should I make an issue + TODO comment?

mabelzhang replied:
Ah, sorry I did not look at the method body closely for `/list`. Looks like Ignition doesn't have the capability to publish an array of markers.

However, @chapulina has suggested using lines and points, instead of cylinders and spheres. https://github.com/ignitionrobotics/ign-msgs/blob/6b4430a91e39e66ad6e9c2844c824e36aebc8b48/proto/ignition/msgs/marker.proto#L147-L154

LINE_LIST and POINTS probably do the same as their counterparts in RViz and should give you similar results. They are essentially arrays, but they only need to be published once. You would `push_back` each point into the vector, and only publish once. http://wiki.ros.org/rviz/DisplayTypes/Marker#Line_List_.28LINE_LIST.3D5.29

mcres replied:
> LINE_LIST and POINTS probably do the same as their counterparts in RViz and should give you similar results. They are essentially arrays, but they only need to be published once

I've tried that, and it makes a huge difference. I don't know if that difference comes from transporting only one message or rendering a simpler figure though.

> Looking at what the /marker service does, it shouldn't be very difficult to implement the service we're looking for.
Should I make an issue + TODO comment?

I think that I should do this as well, it could be a nice `good first issue`. Besides, if the performance using cylinders and spheres is the same, we should probably use those because it looks way cooler.

Here's a screenshot with LINE_LIST and POINTS. Btw, distinguishing the points is really difficult, so I'll probably remove those:

![line_list](https://user-images.githubusercontent.com/14854080/89909840-cc54d600-dbef-11ea-92f8-582a2ccca9f5.png)

mabelzhang replied:
Both. The simpler geometry certainly makes a difference, but the reduction in messages as well. That is my experience with RViz. The performance would not be the same with cylinders and spheres, because they are each rendering many faces. Round geometry especially take longer. I think it is actually good to keep it lines and points for this plugin, because that will enable adding many more plugins. I didn't think of that at the beginning but that is where we should have started.

Try making the point size larger. In RViz, the points are rendered as a single square, and as you rotate the view, you can see them better. Don't remove them, because there are cases when normals may be invalid and the points are the only information the user can go by.

Feel free to create an issue, but we don't need it for this plugin.

mcres replied:
> Try making the point size larger

Hmm I don't know if this is currently possible. I've tried the `scale` field, but all it does is changing the pose of the marker, but the size is always the same. The `pose` field also changes the pose of the marker.

mabelzhang replied:
Sounds like a bug, either in Ignition or in the plugin. I looked for where the marker rendering code sets the scale, and it looks very simple, one block handling all types of markers https://github.com/ignitionrobotics/ign-gazebo/blob/master/src/rendering/MarkerManager.cc#L255

Maybe add a printout from the `has_scale` block and `has_pose` block there and see if it's getting the values you input? Then you can go from there to see where the bug is. If you find the bug is in Ignition, we should open an issue.

mcres replied:
> Maybe add a printout from the has_scale block and has_pose block there and see if it's getting the values you input? 

The data passed is the same that I set. I couldn't find the issue though, but it must be in `ign-rendering`.
The rest of these comment is addressed in e216bb2

mcres commented on src/systems/optical_tactile_plugin/Visualization.cc:
```diff
@@ -0,0 +1,239 @@
+  // <visualization_resolution> if <visualize_forces> is set to true
+
+  // Check if _normalForce has to be visualized
+  bool visualizeForce = (_i % this->visualizationResolution == 0 &&
```
> Instead of doing modulo at all, you can simply increment the i and j loops by visualizationResolution. That should eliminate all of the modulo operations and would not need any conditional checks.

I like this one, but regarding normal forces we have computation (which will be published in a message in the future) and visualization. Does it make sense to not computing the normal forces for the whole range of the depth camera? And if it does, should that resolution be the same of the visualization?

mabelzhang replied:
Do you mean the normals represented as an RGB image? Yes, I think we will need to sacrifice the resolution at the benefit of performance. It should also be more intuitive to the user, that the resolution they specify applies to all the resolutions.

For the other messages (ones that are not an RGB image), the resolution of the normals is independent of the original depth resolution. The normal computation is only using the raw depth values, right? At that, it is only using the immediate neighboring pixels, with indices changing at a magnitude of 1 (i.e. i +/- 1, j +/- 1). So it shouldn't matter what resolution the result is, the computation should be the same.

mcres replied:
> So it shouldn't matter what resolution the normals are computed on, the pixel operations on the raw depth values will be the same.

Correct, just asked in case we wouldn't want different visualization & computation resolutions.

mcres replied:
e216bb2

mcres commented on src/systems/optical_tactile_plugin/Visualization.hh:
```diff
@@ -0,0 +1,125 @@
+      ignition::math::Pose3f &_sensorWorldPose);
+
+    /// \brief Message for visualizing the sensor
+    private: ignition::msgs::Marker sensorMarkerMsg;
```
> It can introduce error, in that if a marker message is configured a certain way, and is reused for another marker, some of the information might not be overwritten and carry to the second marker.

Hmm, I see your point. However, we've seen the plugin's performance is not that good when it comes to visualization, so continuously creating and deleting will make things worse (significantly?). Besides, I can't think why somebody would overwrite the messages variables to use them for something else. These variables are documented and the structure of the `OpticalTactilePluginVisualization` class makes sense to me (I'll admit I'm probably biased there!)

mabelzhang replied:
We will keep this comment on hold, until we figure out whether the markers can be published in a list. It is a big hit on performance, and it is not an option - it's a must.

Reusing the variable is not a deliberate act, it is usually an unintended bug, which can happen in future maintenance and additions. Usually, variables that don't need to be class scope are declared at a local scope just as good practice.

mabelzhang replied:
Following up on https://github.com/ignitionrobotics/ign-gazebo/pull/229#discussion_r468278088 , if that improves the performance a lot, we can demote this to a local variable.

mcres replied:
> Reusing the variable is not a deliberate act, it is usually an unintended bug, which can happen in future maintenance and additions. Usually, variables that don't need to be class scope are declared at a local scope just as good practice.

I understand. As seen in https://github.com/ignitionrobotics/ign-gazebo/pull/229#discussion_r468634575, publishing a list improves the performance, so I'm demoting to a local variable. Should I do this for the normal forces _and_ the sensor marker? They are published each time the depth camera publishes an image and in each `PostUpdate` call, respectively.

mabelzhang replied:
I was thinking definitely `positionMarkerMsg` and `forceMarkerMsg`, which are the arrays I believe. The sensor can be kept as a box. I think it is a good idea to keep it local as well.

mcres replied:
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+    // Get the Depth Camera pose
+    ignition::math::Pose3d sensorPose =
+      _ecm.Component<components::Pose>(
+        this->dataPtr->model.Entity())->Data();
```
> The comment says depth camera pose, but this is actually getting the model pose

Yes, I was actually referring to the model pose.

> I think the frame name is just the model's name, but to be sure, you'll want to test this

I thought the same, but I've actually tested removing the attribute, setting it to `ground_plane` and even names not related to simulation and it seems the plugin keeps doing OK... weird. I think that an `ignwarn` if the model's name and `relative_to` value are not equal is still a good idea.

mabelzhang replied:
Huh! That is weird. I'm seeing sdformat tests that in `ign_TEST.cc` https://github.com/osrf/sdformat/blob/master/src/ign_TEST.cc#L499

Looks like it wants link, joint, or frame name. So probably use the link name instead of model name for the `ignwarn` check.

azeey replied:
The `relative_to` attribute of poses for sensors is resolved during [creation]( https://github.com/ignitionrobotics/ign-gazebo/blob/fd37a777122e214b8051ea404eaaa151d110b52e/src/SdfEntityCreator.cc#L519) and it is expressed relative to the parent element (link or joint) that contains the sensor, so there is no need to handle the `relative_to` attribute here. This is true for all other SDFormat elements as well so long as their pose is defined in the SDFormat spec. i.e, this doesn't include `<pose>` elements inside `<plugin>` for example.

mabelzhang replied:
Cool, thanks for the clarification!

mcres replied:
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+    // We call Load here instead of Configure because we can't be guaranteed
+    // that all entities have been created when Configure is called
+    this->dataPtr->Load(_ecm);
+    this->dataPtr->initialized = true;
```
In that case, if it fails to initialize it will continuously try to `Load()` and printing an error/warning many times per second.
On the other hand, if `initialize` is set like it currently is, it will print the error/warning once and will continue running, probably failing or doing weird things.

Another option is using another flag, e.g. `triedToLoad` to make sure we only `Load()` once. That way, if the plugin fails to load, the error/warning will only be printed once, and after that the plugin will do nothing.

mabelzhang replied:
It's typical practice to have a class Boolean variable e.g. `initErrorPrinted`, so that it only prints once. In the printout, you can add in parentheses `(only printed once)`. You don't need to limit how many times `Load()` is called, because something could have changed dynamically that make later calls to `Load()` succeed, e.g. someone injects into the SDF or changes the components in the ECM.

mcres replied:
e216bb2

mcres commented on src/systems/optical_tactile_plugin/Visualization.cc:
```diff
@@ -0,0 +1,239 @@
+
+//////////////////////////////////////////////////
+OpticalTactilePluginVisualization::OpticalTactilePluginVisualization(
+  double &_contactRadius, double &_forceRadius, double &_forceLength,
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
 
-  // Variable for setting the markers id through the iteration
-  int marker_id = 1;
+  // todo(anyone) Improve performance once the following issue is completed:
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/Visualization.cc:
```diff
@@ -0,0 +1,239 @@
+  this->forceMarkerMsg.set_id(this->markerID++);
+  this->forceMarkerMsg.set_action(ignition::msgs::Marker::ADD_MODIFY);
+  this->forceMarkerMsg.mutable_lifetime()->set_sec(this->cameraUpdateRate);
+  this->forceMarkerMsg.mutable_lifetime()->set_nsec(0);
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/Visualization.cc:
```diff
@@ -0,0 +1,239 @@
+  this->sensorMarkerMsg.set_visibility(ignition::msgs::Marker::GUI);
+
+  // Set material properties
+  this->positionMarkerMsg.mutable_material()->mutable_ambient()->set_r(0);
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/Visualization.hh:
```diff
@@ -0,0 +1,125 @@
+    private: int visualizationResolution;
+
+    /// \brief Variable for setting the markers id
+    private: int markerID;
```
e216bb2

I deleted the variable, as now I'm using one marker only (a list of points and another one lines).

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.hh:
```diff
@@ -45,23 +46,30 @@ namespace systems
-    /// value is 100.
+    /// <visualization_resolution> Number of pixels to skip when visualizing
+    /// the forces. One vector representing a normal force is computed for each
+    /// of the camera pixels. This element must be positive and it is optional.
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -43,39 +43,42 @@
+
+  /// \brief Map the (i,j) coordinates defined in the top-left corner of the
+  /// camera into its corresponding (X,Y,Z) measurement with respect to the
+  /// camera's origin
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -43,39 +43,42 @@
+  /// \param[in] _contact Message containing the data to interpolate
+  public: void InterpolateContactData(const ignition::msgs::Contact &_contact);
 
   /// \brief Callback for the Depth Camera
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -43,39 +43,42 @@
+
+  /// \brief Check if a specific point returned by the Depth Camera is inside
+  /// the contact surface.
+  /// \param[in] _point Point returned by the Depth Camera
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -43,39 +43,42 @@
+  public: bool PointInsideSensor(ignition::math::Vector3f _point);
 
   /// \brief Computes the normal forces of the Optical Tactile sensor
   /// \param[in] _msg Message returned by the Depth Camera
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -159,8 +150,8 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
 
-  /// \brief 3D vector to store unpacked XYZ points
-  std::vector<std::vector<std::vector<float>>> imageXYZ;
+  /// \brief Data returned by the Depth Camera after being preprocessed
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
 
-    // Get sdf parameters
-    if (!_sdf->HasElement("resolution"))
+  // <enabled>
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+  this->dataPtr->sensorSize =
+    _sdf->GetParent()->GetElement("link")->GetElement("collision")->
+    GetElement("geometry")->GetElement("box")->
+    Get<ignition::math::Vector3d>("size");
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+  // Check plugin structure
+  // Model should only have 1 link
+  auto allLinks =
+    _ecm.ChildrenByComponents(this->model.Entity(), components::Link());
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
 
-    // Clear vector that contains old data
-    this->filteredCollisionEntities.clear();
+  auto* contacts =
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+    this->visualizationResolution);
+
+  // Create the marker messages
+  this->visualizePtr->CreateMarkersMsgs(this->modelName, this->sensorSize);
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+  // Link should have a collision component linked to a contact sensor
+  auto linkCollisions =
+    _ecm.ChildrenByComponents(allLinks[0], components::Collision());
+  for (const Entity &colEntity : linkCollisions)
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+  }
 
-    if (!_sdf->HasElement("force_length"))
+  // Link should have exactly 1 Contact sensor and 1 Depth Camera sensor
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
-}
+  if ((depthCameraCounter != 1) || (contactSensorCounter != 1))
+  {
+    ignerr << "Link must have 1 Depth Camera sensor and "
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+  // Configure subscriber for Depth Camera images
+  if (!depthCameraSdf->HasElement("topic"))
+  {
+    ignerr << "Depth camera should have a unique <topic> value" << std::endl;
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
-    if (_entities.empty())
-        return;
+  // Nothing left to do if paused.
+  if (_info.paused)
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+void OpticalTactilePluginPrivate::InterpolateContactData(
+  const ignition::msgs::Contact &_contact)
+{
+  // The Optical Tactile Sensor is supposed to be box-shaped.
```
Removed method: #229 (comment)

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -170,547 +161,472 @@ class ignition::gazebo::systems::OpticalTactilePluginPrivate
+  const ignition::msgs::Contact &_contact)
+{
+  // The Optical Tactile Sensor is supposed to be box-shaped.
+  // Contacts returned by the Contact Sensor in this type of shapes should be
```
Removed method: #229 (comment)

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
+  ignition::math::Vector3f measuredPoint;
+
+  // Save original position of the pointer
+  char *initialMsgBuffer = this->msgBuffer;
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
+bool OpticalTactilePluginPrivate::PointInsideSensor(
+  ignition::math::Vector3f _point)
+{
+  // We assume that the Depth Camera is only displaced in the -X direction
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
+  bool insideX = (_point.X() >= std::abs(
+    this->depthCameraOffset.X()) - this->extendedSensing)
+    && (_point.X() <= (std::abs(this->depthCameraOffset.X()) +
+    this->sensorSize.X() + this->extendedSensing));
```
e216bb2

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -720,208 +636,123 @@ void OpticalTactilePluginPrivate::DepthCameraCallback(
+  bool insideY = (_point.Y() <= this->sensorSize.Y() / 2 +
+    this->extendedSensing)
+    && (_point.Y() >= - this->sensorSize.Y() / 2 -
+    this->extendedSensing);
```
e216bb2

mcres commented:
I've addressed the comments and it should be ready for another review.
Apart from the minor changes, there were a few comments that made do some more refactoring:

- In order to change the rendering to cylinders&spheres to points&lines, along with making the marker msg variables local, I had to refactor `Visualization.hh` and `Visualization.cc`. I've basically divided those methods into initialization (setting some common msg fields), addition in the case normal forces and request of the `/marker` service. I hope it's clearer than before.
- Given the marker types change, I've also removed the plugin parameters that are no longer necessary.
- I've removed the `InterpolateContactData` method.

mabelzhang commented:
Looks good. The only remaining thing to look into is the POINTS marker not taking on the scales specified, but that can be done outside of this PR.

DCO looks peculiar... probably from one of the merged commits. I might need to override that.

@azeey Could you take a last pass since you requested changes? Also good for another pair of eyes since we're merging into `master`.

mabelzhang commented:
@azeey Friendly ping for a review please :)

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,702 @@
+  }
+
+  // Store depth camera update rate and offset from model
+  auto offset = depthCameraSdf->Get<ignition::math::Pose3d>("pose");
```
Thanks!

mcres commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,702 @@
+    return;
+
+  // Get data from the message
+  std::string rawMsgBuffer = _msg.data();
```
Ahh, silly me, I didn't realize I could still modify the pointer that way. Thanks!

mabelzhang commented:
For visibility, we are reviewing newer additions into this branch on @mcres's fork https://github.com/mcres/ign-gazebo/pull/1 for the sake of time, so that we can get everything in by this Friday.

When this PR here is merged, we will open a new PR with changes from that fork. Hopefully that will take shorter time to review since I would have done a few passes.

mabelzhang commented:
Could you merge from master? That might fix the 3 new warnings on CI.

azeey commented on src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc:
```diff
@@ -0,0 +1,702 @@
+  {
+    // We call Load here instead of Configure because we can't be guaranteed
+    // that all entities have been created when Configure is called
+    this->dataPtr->Load(_ecm);
```
My suggestion was just for `Load`. `components::ContactSensorData` may not be available during `Configure` because the `Contact` system creates it in `PreUpdate`. Either way, I suggest checking if `csd` is null before doing anything with it because it's not guaranteed to exist.
https://github.com/ignitionrobotics/ign-gazebo/blob/b321a0044e4feab3e6a1406139691581b90ad892/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L326  This would be the case, for example, if the `Contact` system was not added to the SDF file.

azeey replied:
The added benefit of calling `Load` in `Configure` is you don't have to check `initErrorPrinted`. `Configure` is called only once.

mabelzhang replied:
@mcres Friendly ping. Do you have time to address the comment above? Looks like that is the last comment (and the previous Mac compiler error on CI, which is rerunning now), unless @azeey has more to review.
Let us know if you want us to help with the branch. Ideally, we'd like to get this in before code freeze on Sep 23.

mabelzhang replied:
Okay I see what's going on here. @azeey So I tried moving the `Load()` call into `Configure()`, and as @mcres mentioned, the `csd` pointer in `PreUpdate()` is then just always null. I left it running for a while, and it never went non-null.

The same approach is used in the `TouchPlugin`, to call `Load()` from `PreUpdate()` instead of `Configure()`. I guess it must be similar here?
https://github.com/ignitionrobotics/ign-gazebo/blob/main/src/systems/touch_plugin/TouchPlugin.cc#L368-L374

azeey replied:
Ah, the issue is that we're checking for `ContactSensorData` in `Load` to determine the Collision associated with a contact sensor. That component won't be available until the `Contact` system has ran once.
https://github.com/ignitionrobotics/ign-gazebo/blob/b17e904d5d7884aa592ce3e4f38019f03f866bb1/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L402-L403

So, yeah, you guys are correct. `Load` needs to be in `PreUpdate`. And since currently, world systems are ran before model systems, the `Contact` system is guaranteed to run before the `OpticalTactilePlugin`.

mabelzhang replied:
Cool! Thanks @azeey for looking into it. Could you approve the PR? I think it won't let us merge if there are remaining "Requested changes" marks. I made a PR to the fork to fix the Mac warning.

chapulina commented:
I'm going to remove the `beta` label from this PR, so it won't be included on Dome's initial release, but it can come on a minor release after that.

mcres commented:
Seems that the Ubuntu CI was fixed. Should I do anything else before merging?

chapulina commented:
@mabelzhang / @azeey : both of you have requests for changes. Would it be possible to do another review? Thanks!

mabelzhang commented:
Looks like this comment is unresolved: https://github.com/ignitionrobotics/ign-gazebo/pull/229#discussion_r482524013
Mac CI is still failing.
Those are the only remaining things for this PR.

mabelzhang commented:
@osrf-jenkins run tests

mabelzhang commented:
Mac CI has a new warning https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-homebrew-amd64/3994/
```
OpticalTactilePlugin.cc:529
expression result unused
```
That should be an easy fix.

Test failures:
ServerRepeat/DiffDriveTest.PublishCmdCustomTopics/0
ServerRepeat/DiffDriveTest.SkidPublishCmd/0
INTEGRATION_physics_system.test_ran

@chapulina are these flakey by any chance? The reference build 3993 has a different set of failing tests, so I can't really tell. Looks like the badge linked from osrf/buildfarmer has yet a different set of failing tests ha...

chapulina commented:
> are these flakey by any chance?

We don't have many builds on `ign-gazebo4` yet, but luckily we have enough to see these 2 are flaking:

* https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo4-homebrew-amd64/3/testReport/junit/(root)/ServerRepeat_DiffDriveTest/PublishCmdCustomTopics_0/history/
* https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo4-homebrew-amd64/3/testReport/junit/(root)/ServerRepeat_DiffDriveTest/SkidPublishCmd_0/history/


But the `NonDefaultCanonicalLink` test that segfaulted has been passing on `ign-gazebo4`:

* https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo4-homebrew-amd64/3/testReport/junit/(root)/PhysicsSystemFixture/NonDefaultCanonicalLink/history/

And `ign-gazebo3`:

* https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo3-homebrew-amd64/68/testReport/junit/(root)/PhysicsSystemFixture/NonDefaultCanonicalLink/history/

It may be worth rerunning just to see if it wasn't a spurious failure.

mabelzhang commented:
Ah, cool, took me a few minutes to figure out the correct order to click on History to get to those pages you linked. Nice graphs. Jenkins could use better UI.

Ubuntu CI has one failure, 
`ContactSystemTest.MultipleCollisionsAsContactSensors`, but looks like it's failed once before
https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/4316/testReport/(root)/ContactSystemTest/MultipleCollisionsAsContactSensors/history/

mabelzhang commented:
@osrf-jenkins run tests

mabelzhang commented:
Grabbing the results before Jenkins shuts down today.

Ubuntu CI:
https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/4319/
Test Result (2 failures / +1)
ContactSystemTest.MultipleCollisionsAsContactSensors
EntityComponentManagerPerfrormance.Each

I am not sure if the ContactSystem test came from this PR... I saw another failure before 4316, and I think the earlier one is not from this PR. I didn't write down the number.
```
/var/lib/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/ign-gazebo/test/integration/contact_system.cc:118
Value of: contactMsgs.size()
  Actual: 3
Expected: 0u
Which is: 0
```

Mac CI:
https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-homebrew-amd64/3997/
There's a seg fault still.
Test Result (2 failures / +1)
PeerTracker.PeerTrackerStale
UNIT_Server_TEST.test_ran

So looks like the ones from 3994 above were flaky.
These new ones did not fail in 3994. PeerTrackerStale failed for both 3997 and 3998. So these look flaky.

azeey commented on src/systems/optical_tactile_plugin/Visualization.cc:
```diff
@@ -0,0 +1,212 @@
+    _normalForce * this->forceLength, normalForceOrientationFromSensor);
+
+  normalForcePoseFromWorld =
+    (normalForcePoseFromSensor + this->depthCameraOffset) + _sensorWorldPose;
```
Consider using `Pose3f::operator*` instead of `Pose3f::operator+` since the latter is [discouraged](https://github.com/ignitionrobotics/ign-math/issues/60). If you do, note that you'll have to swap the order of operations

azeey commented on src/systems/optical_tactile_plugin/Visualization.cc:
```diff
@@ -0,0 +1,212 @@
+
+//////////////////////////////////////////////////
+OpticalTactilePluginVisualization::OpticalTactilePluginVisualization(
+  std::string &_modelName,
```
Should these be const refs?

azeey reviewed (approved):
LGTM! Left just a couple of minor comments. Feel free to address them in the follow up PR.

mabelzhang commented:
[ignition_gazebo-ci-pr_any-homebrew-amd64](https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-homebrew-amd64/4233/) is still reporting some failures:
```
The following tests FAILED:
	 79 - INTEGRATION_joint_state_publisher_system (SEGFAULT)
	 83 - INTEGRATION_level_manager (Subprocess aborted)
```

The first one I don't see anywhere else.

Related to the second one, level manager had some related failures, on
`LevelManagerPerfrormance.LevelVsNoLevel` https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo4-homebrew-amd64/4/
and `INTEGRATION_level_manager_runtime_performers.test_ran` https://build.osrfoundation.org/job/ignition_gazebo-ci-ign-gazebo4-homebrew-amd64/3/

[ignition_gazebo-ci-pr_any-ubuntu_auto-amd64](https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/4745/)'s 2 out of 3 failures are expected (`LogSystemTest.LogResources` and `SdfInclude.DownloadFromFuel`).
This one failed but shouldn't: `ServerRepeat/VelocityControlTest.PublishCmd/0`

However, seeing that all the code changes in this PR are strictly within a new directory `src/systems/optical_tactile_plugin`, I really don't think it should affect any of the tests at all...
Will rerun and see if these fail again.

mabelzhang commented:
@osrf-jenkins run tests

mabelzhang commented:
Alright, those two on homebrew-amd64 were flaky. Not failing on the rerun. Rerun fails on INTEGRATION_breadcrumbs which is flaky.

ubuntu_auto-amd64 still failing on ServerRepeat/VelocityControlTest.PublishCmd/0, which is unexpected.
Looks like it could be a precision problem, but it's not comparing `>=`, it wants `>`.
```
123: /var/lib/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/ign-gazebo/test/integration/velocity_control_system.cc:128: Failure
123: Expected: (poses[i].Pos().X()) > (poses[i-1].Pos().X()), actual: 2.12913e-10 vs 2.13556e-10
123: /var/lib/jenkins/workspace/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/ign-gazebo/test/integration/velocity_control_system.cc:129: Failure
123: Expected: (poses[i].Pos().Y()) > (poses[i-1].Pos().Y()), actual: 2 vs 2
```

However, the previous run had 4 unexpected comparison outcomes, the latest one only has 2. So it could just be flaky.
I will rerun one more time to be careful.

mabelzhang commented:
@osrf-jenkins run tests

mabelzhang commented:
It looks like ServerRepeat/VelocityControlTest.PublishCmd/0 on ubuntu_auto-amd64 is really failing just for this PR. All other tests are passing. Need to look into it.

mabelzhang commented:
Huh, that test is failing elsewhere too
https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/4751/testReport/
https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/4752/testReport/
https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/4755/testReport/

I think it's flaky.
That is the last test failure. I'm ready to approve this. Might wait for a few more runs on this test to see if it fails more to be sure.

mabelzhang commented:
I looked again, the 3 failures of the same test above come from 3 different PRs, so it's definitely flaky. Merging this.

adlarkin commented:
@mabelzhang was this just a regular merge? It looks like we forgot to squash the PR into one commit before merging. Here's the diff between `ign-gazebo4` and a branch I created just before thus was merged: https://github.com/ignitionrobotics/ign-gazebo/compare/adlarkin/4.1.0...ign-gazebo4

mabelzhang commented:
Yes just a regular merge, sorry I wasn't sure what was the preferred method of merge and just went with the default. Noted for the future.
